### PR TITLE
[#IOPID-1953, IOPID-1954, IOPID-1955] creating fn-public in ITN

### PR DESCRIPTION
**Pull request by arcogabbo**
### Major Changes
* [[#IOPID-1953] created fn-public in ITN](https://github.com/pagopa/io-infra/commit/9b90a87388d1764f33568f1706218b821e6e99d2)

### Dependencies
None

### Testing
None

### CODE REVIEW

File: src/domains/citizen-auth-app/01_network_itn.tf [hunk 1]
```diff
@@ -246,3 +246,50 @@ resource "azurerm_private_endpoint" "staging_function_profile_async_itn_sites" {
...
+    name                 = "private-dns-zone-group"
+    private_dns_zone_ids = [data.azurerm_private_dns_zone.privatelink_azurewebsites_net.id]
+  }
+
+  depends_on = [module.function_public_itn]
+
+  tags = var.tags
+}
+
+resource "azurerm_private_endpoint" "staging_function_public_itn_sites" {
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/01_network_itn.tf:274]: Was this one missing?

File: src/domains/citizen-auth-app/01_network_itn.tf [hunk 2]
```diff
@@ -246,3 +246,50 @@ resource "azurerm_private_endpoint" "staging_function_profile_async_itn_sites" {
 
   tags = var.tags
 }
+
+## fn-public-itn pep
+resource "azurerm_private_endpoint" "function_public_itn_sites" {
+  name                = "${local.common_project_itn}-public-fn-pep-01"
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/01_network_itn.tf:252]:
```suggestion
  name                = "${local.common_project_itn}-public-func-pep-01"
```

File: src/domains/citizen-auth-app/01_network_itn.tf [hunk 3]
```diff
@@ -246,3 +246,50 @@ resource "azurerm_private_endpoint" "staging_function_profile_async_itn_sites" {
...
+  location            = local.itn_location
+  resource_group_name = azurerm_resource_group.public_rg_itn.name
+  subnet_id           = data.azurerm_subnet.itn_pep.id
+
+  private_service_connection {
+    name                           = "${local.common_project_itn}-public-fn-pep-01"
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/01_network_itn.tf:258]:
```suggestion
    name                           = "${local.common_project_itn}-public-func-pep-01"
```

File: src/domains/citizen-auth-app/01_network_itn.tf [hunk 4]
```diff
@@ -246,3 +246,50 @@ resource "azurerm_private_endpoint" "staging_function_profile_async_itn_sites" {
...
+    private_dns_zone_ids = [data.azurerm_private_dns_zone.privatelink_azurewebsites_net.id]
+  }
+
+  depends_on = [module.function_public_itn]
+
+  tags = var.tags
+}
+
+resource "azurerm_private_endpoint" "staging_function_public_itn_sites" {
+  name                = "${local.common_project_itn}-fast-login-fn-staging-pep-01"
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/01_network_itn.tf:275]:
```suggestion
  name                = "${local.common_project_itn}-fast-login-func-staging-pep-01"
```

File: src/domains/citizen-auth-app/01_network_itn.tf [hunk 5]
```diff
@@ -246,3 +246,50 @@ resource "azurerm_private_endpoint" "staging_function_profile_async_itn_sites" {
...
+  location            = local.itn_location
+  resource_group_name = azurerm_resource_group.public_rg_itn.name
+  subnet_id           = data.azurerm_subnet.itn_pep.id
+
+  private_service_connection {
+    name                           = "${local.common_project_itn}-public-fn-staging-pep-01"
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/01_network_itn.tf:281]:
```suggestion
    name                           = "${local.common_project_itn}-public-func-staging-pep-01"
```

File: src/domains/citizen-auth-app/10_function_public.tf [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+module "function_public_snet_itn" {
+  source                                    = "git::https://github.com/pagopa/terraform-azurerm-v3.git//subnet?ref=v8.44.0"
+  name                                      = format("%s-public-snet-01", local.short_project_itn)
+  address_prefixes                          = var.cidr_subnet_fnpublic_itn
+  resource_group_name                       = data.azurerm_virtual_network.common_vnet_italy_north.resource_group_name
+  virtual_network_name                      = data.azurerm_virtual_network.common_vnet_italy_north.name
+  private_endpoint_network_policies_enabled = true
+
+  service_endpoints = [
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/10_function_public.tf:51]: What are the downstream dependencies of this function app? We can remove Cosmos DB for sure, maybe the others as well

File: src/domains/citizen-auth-app/10_function_public.tf [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    "private_dns_zone_queue_ids" = [data.azurerm_private_dns_zone.privatelink_queue_core_windows_net.id],
+    "private_dns_zone_table_ids" = [data.azurerm_private_dns_zone.privatelink_table_core_windows_net.id],
+    "queues"                     = [],
+    "containers"                 = [],
+    "blobs_retention_days"       = 0,
+  }
+
+  subnet_id = module.function_public_snet_itn.id
+
+  allowed_subnets = [
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/10_function_public.tf:106]: We've been using peps, so this property should be null and `enable_function_app_public_network_access` must be set to `false`.

However, as we've seen last time this condition is not true when DevOps is being used. If this is the case, set `data.azurerm_subnet.apim_v2_snet.id` as the only allowed subnet and do not set `enable_function_app_public_network_access` variable

File: src/domains/citizen-auth-app/10_function_public.tf [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+  runtime_version                          = "~4"
+  application_insights_instrumentation_key = data.azurerm_application_insights.application_insights.instrumentation_key
+
+  app_settings = merge(
+    local.function_public.app_settings_common,
+  )
+
+  subnet_id = module.function_public_snet_itn.id
+
+  allowed_subnets = [
```
Comment (Krusty93) [inline src/domains/citizen-auth-app/10_function_public.tf:146]: Same as above

**Review by Krusty93 — CHANGES_REQUESTED** [review]

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 7 above)
Comment (arcogabbo) [reply src/domains/citizen-auth-app/10_function_public.tf:106]: thank you @Krusty93. we are still using azdevops here, I removed the unnecessary in https://github.com/pagopa/io-infra/pull/1244/commits/8feb975ccc66152b87932e21eb9a2ef89b5ee2eb

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 8 above)
Comment (arcogabbo) [reply src/domains/citizen-auth-app/10_function_public.tf:145]: I removed the unnecessary in https://github.com/pagopa/io-infra/pull/1244/commits/8feb975ccc66152b87932e21eb9a2ef89b5ee2eb

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 6 above)
Comment (arcogabbo) [reply src/domains/citizen-auth-app/10_function_public.tf:51]: cosmos is used to create a profile model and retrieve some data. storage is used to retrieve data from a table storage. I believe we need to keep them

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 7 above)
Comment (gquadrati) [reply src/domains/citizen-auth-app/10_function_public.tf:106]: I think we can safely remove the Apim submit, anyway. 
The only one we still need is `data.azurerm_subnet.azdoa_snet[0].id` for staging.

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 6 above)
Comment (Krusty93) [reply src/domains/citizen-auth-app/10_function_public.tf:51]: If downstream resources use private endpoints, we don't need to set the service endpoints (all Cosmos instances use peps as far as I know). Please check if the storage account provides a private endpoint for the table endpoint

File: src/domains/citizen-auth-app/10_function_public.tf (hunk 7 above)
Comment (Krusty93) [reply src/domains/citizen-auth-app/10_function_public.tf:106]: Great

Comment (arcogabbo) [reply src/domains/citizen-auth-app/10_function_public.tf:100]: done in https://github.com/pagopa/io-infra/pull/1244/commits/020325834ce684b2025231e96902ff9e247579ea

**Review by gquadrati — APPROVED** [review]
LGTM

**Review by BurnedMarshal — APPROVED** [review]

**Pull request merged**